### PR TITLE
Load the main JS bundle in production with async

### PR DESCRIPTION
**Pull request by arunoda**
Fixes https://github.com/zeit/next.js/issues/1436

With this, we don't need to wait for the JS bundle and page can render directly.
This will lead to better page load performance.

### CODE REVIEW

**Comment by ptomasroos** [comment]
Great improvments @arunoda

**Review by timneutkens — APPROVED** [review]
Niiiiice 😍 ❤️

File: server/build/index.js [hunk 1]
```diff
@@ -44,17 +45,18 @@ function runCompiler (compiler) {
...
-  webpackStats.chunks
-    // We are not interested about pages
-    .filter(({ files }) => !/^bundles/.test(files[0]))
-    .forEach(({ hash, files }) => {
-      chunkHashMap[files[0]] = { hash }
-    })
-
+async function writeBuildStats (dir) {
+  // Here we can't use the hashes in the webpack chunks.
+  // That's because the "app.js" is not tight with a chunk.
```
Comment (timneutkens) [inline server/build/index.js]: > is not tight with a chunk.

`is not tied to a chunk` is more correct here right?

File: server/build/plugins/combine-assets-plugin.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
+// This pulgin combines a set of assets into a single asset
```
Comment (timneutkens) [inline server/build/plugins/combine-assets-plugin.js]: `pulgin` => `plugin`

File: server/build/index.js (hunk 1 above)
Comment (arunoda) [inline server/build/index.js]: Thanks.

**Comment by matthewmueller** [comment]
should we also defer this bundle?

**Comment by arunoda** [comment]
> should we also defer this bundle?
No. Defer and Async are different stuff. We might use that for backward compatibility.
But since we only focus on IE9+, this is fine.

**Comment by tomevans18** [comment]
I feel like this topic has been done to death but I wanted to comment and make sure I'm on the right lines and nothing has been missed.

After trying to improve the performance of a site I'm working on I came across this PR.
On review of Chrome's performance profiler, I noticed that the main.js file is taking a long time to evaluate (1.6 seconds in my case) and this seems to be slowing down the sites first meaningful paint.

My initial thought is deferring this file (along with all other JS) as it is not needed initially due to SSR.

image

What are your thoughts? Maybe all JS should be loaded after the DOM on initial load.
Is this achievable?

**Comment by timneutkens** [comment]
The long evaluation is probably fixed by https://github.com/zeit/next.js/pull/4639 as main.js used to serve as main bundle + commons.

In general, it makes not much sense to comment on a 1-year-old pull request to ask for help.

**Comment by tomevans18** [comment]
@timneutkens thanks for the info regarding #4639, I will keep an eye on that PR.

Apologies, I understood it was an old PR hence cavitating the comment with "I feel like this topic has been done to death". I could see that there were several comments on an issue (#1436) and this PR was to resolve it, with no future issues or PRs referencing this problem... hence the comment above.

**Pull request merged**